core/services/ocr2/delegate.go

```diff
@@ -343,23 +347,95 @@ func (d Delegate) ServicesForSpec(jobSpec job.Job) ([]job.ServiceCtx, error) {
 			runResults,
 			d.pipelineRunner,
 			make(chan struct{}),
-			lggr)
+			lggr,
+		)
 
 		// NOTE: we return from here with the services because the OCR2VRF oracles are defined
 		// and exported from the ocr2vrf library. It takes care of running the DKG and OCR2VRF
 		// oracles under the hood together.
 		oracleCtx := job.NewServiceAdapter(oracles)
 		return []job.ServiceCtx{runResultSaver, vrfProvider, oracleCtx}, nil
+	case job.OCR2Keeper:
```

Comment:
Everything in this case, `job.OCR2Keeper` should be broken out to its own function and paired with a unit test. There is a lot of code doing a lot of magic here. This section of code is touched by many people across multiple teams that don't know the expectations of each product setup. One accidental change here could snowball and be hard to detect. It's best to set expectations firmly by applying a unit test to the scope of what is important to us.